cohort-engine/src/main/java/com/ibm/cohort/engine/measure/CDMMeasureEvaluation.java

```diff
@@ -0,0 +1,197 @@
+	 * Evaluate a CDM Patient Quality Measure
+	 * 
+	 * @param measure   CDM Patient Quality Measure
+	 * @param context   CQL Engine Execution Context pre-configured for use in
+	 *                  measure evaluation
+	 * @param patientId Patient ID of the patient to evaluate
+	 * @return MeasureReport with population components filled out.
+	 */
+	public MeasureReport evaluatePatientMeasure(Measure measure, Context context, String patientId) {
+		MeasureReport report = evaluation.evaluatePatientMeasure(measure, context, patientId);
+		;
+
+		MeasureScoring scoring = MeasureScoring.fromCode(measure.getScoring().getCodingFirstRep().getCode());
+		switch (scoring) {
+		case PROPORTION:
+		case RATIO:
+			// implement custom logic for CDM care-gaps
+			Iterator<Measure.MeasureGroupComponent> it = measure.getGroup().iterator();
+			for (int i = 0; it.hasNext(); i++) {
+				Measure.MeasureGroupComponent group = it.next();
+				MeasureReport.MeasureReportGroupComponent reportGroup = report.getGroup().get(i);
+				boolean evaluateCareGaps = isEligibleForCareGapEvaluation(reportGroup);
+
+				for (Measure.MeasureGroupPopulationComponent pop : group.getPopulation()) {
+					if (pop.getCode().hasCoding(CDM_CODE_SYSTEM_MEASURE_POPULATION_TYPE, CARE_GAP)) {
+						Boolean result = Boolean.FALSE;
+						if (evaluateCareGaps) {
+							result = evaluateCriteria(context, pop.getCriteria().getExpression());
+						}
+
+						MeasureReport.MeasureReportGroupPopulationComponent output = new MeasureReport.MeasureReportGroupPopulationComponent();
+						output.setId(pop.getId()); // need this to differentiate between multiple instances of care-gap
+						output.setCode(pop.getCode());
+						output.setCount(result ? 1 : 0);
+						reportGroup.addPopulation(output);
+					}
+				}
+			}
+			break;
+		default:
+			// no customizations needed
+		}
+
+		return report;
+	}
+
+	/**
+	 * Given the results in a report group determine whether or not the patient is
+	 * eligible for care gap evaluation. Care gaps are applied after all the normal
+	 * report logic with all of the normal report rules about initial-population,
+	 * numerator, and denominator.
+	 * 
+	 * @param reportGroup Report group containing population results for standard
+	 *                    patient quality measure reporting.
+	 * @return true when care gaps should be evaluated, otherwise false.
+	 */
+	private boolean isEligibleForCareGapEvaluation(MeasureReport.MeasureReportGroupComponent reportGroup) {
+		boolean isEligibleForCareGap = false;
+		StandardReportResults results = StandardReportResults.fromMeasureReportGroup(reportGroup);
+		// Logic for the numerator exclusion, denominator exclusion, and denominator
+		// exception has already been applied by the standard report generator and the
+		// patient has been removed from the following populations as needed, so we can
+		// keep this logic simple.
+		if (results.inInitialPopulation() && results.inDenominator() && !results.inNumerator()) {
```

Comment:
I think we might need to expand on this logic - given these are logically included and not define meets, we don't need inInitialPopulation() (you have to be to be in denominator).  And I think we'll also need not in numerator exclusion and denominator exception... although I'm somewhat concerned looking at this code https://github.com/DBCG/cqf-ruler/blob/master/r4/src/main/java/org/opencds/cqf/r4/evaluation/MeasureEvaluation.java#L401 that it doesn't account for denominator exclusion or numerator exclusion.

Reply:
The numerator exclusion, denominator exclusion, and denominator exception rules are definitely applied by the cqf-ruler logic. If the patient meets the exclusion criteria, they are removed from the corresponding population (e.g. numerator and numerator exclusion) and added to the exclusion population. See https://github.com/DBCG/cqf-ruler/blob/164e651c3be9a468fac6f6a5be838bd072aaba76/r4/src/main/java/org/opencds/cqf/r4/evaluation/MeasureEvaluation.java#L414 and 
https://github.com/DBCG/cqf-ruler/blob/164e651c3be9a468fac6f6a5be838bd072aaba76/r4/src/main/java/org/opencds/cqf/r4/evaluation/MeasureEvaluation.java#L207

Reply:
Thanks for the pointer... I feel like their code would have been more obvious if it handled all the scenarios the same way.  However, I feel like the first half of my comment is still valid - we don't need to check inInitialPopulation if we're checking inDenominator, and we do need to check for numeratorExclusion and denominatorException.

i.e. if a patient meets the initial population, denominator, and fails the numerator they might then meet the denominator exception criteria.  If they do, I don't think they should be considered for care gap calculation.

Reply:
I agree that the inInitialPopulation check is redundant. I don't agree that we need additional logic for the exclusion and exception criteria. That logic was already applied.  For your example, a patient that met the denominator exception criteria would have already been removed from the denominator population by the cqf-ruler before our logic for caregaps runs. We use the output of cqf-ruler as input to our processing. This is different than if we were checking each of the criteria evaluations ourselves.

Reply:
You are correct about denominator exception, my thinking was wrong.  Would like to see the initial pop check removed but approving regardless.